File: internal/api/tasks.go
package api

import (
	"math"
	"net/http"
	"time"
)

// GET /api/tasks?assignee=&status=
func (rt *Router) handleTasks(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}
	assignee := r.URL.Query().Get("assignee")
	status := r.URL.Query().Get("status")

	tasks, err := rt.Tasks.ListTasks(assignee, status)
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	writeJSON(w, map[string]any{
		"tasks": tasks,
		"count": len(tasks),
	})
}

type assigneeStats struct {
	Assignee string `json:"assignee"`
	Pending  int    `json:"pending"`
	Done     int    `json:"done"`
	Overdue  int    `json:"overdue"`
}

// parseTime tries multiple timestamp formats that SQLite CURRENT_TIMESTAMP may produce.
func parseTime(s string) time.Time {
	for _, fmt := range []string{
		"2006-01-02 15:04:05",
		"2006-01-02T15:04:05Z",
		"2006-01-02T15:04:05",
		"2006-01-02",
	} {
		if t, err := time.Parse(fmt, s); err == nil {
			return t
		}
	}
	return time.Time{}
}

// GET /api/tasks/stats
func (rt *Router) handleTaskStats(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		http.Error(w, "method not allowed", http.StatusMethodNotAllowed)
		return
	}

	all, err := rt.Tasks.ListTasks("", "")
	if err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	now := time.Now()
	today := now.Format("2006-01-02")

	byAssignee := map[string]*assigneeStats{}
	var totalDays float64
	var daysCount int
	var totalDone int
	var recentCompletions []any

	for _, t := range all {
		s, ok := byAssignee[t.Assignee]
		if !ok {
			s = &assigneeStats{Assignee: t.Assignee}
			byAssignee[t.Assignee] = s
		}

		switch t.Status {
		case "pending":
			s.Pending++
			if t.DueDate != "" && t.DueDate < today {
				s.Overdue++
			}
		case "done":
			s.Done++
			totalDone++
			if totalDone <= 10 {
				recentCompletions = append(recentCompletions, t)
			}
			created := parseTime(t.CreatedAt)
			updated := parseTime(t.UpdatedAt)
			if !created.IsZero() && !updated.IsZero() {
				days := updated.Sub(created).Hours() / 24
				if days < 0 {
					days = 0
				}
				totalDays += days
				daysCount++
			}
		}
	}

	stats := make([]assigneeStats, 0, len(byAssignee))
	for _, s := range byAssignee {
		stats = append(stats, *s)
	}

	avgDays := 0.0
	if daysCount > 0 {
		avgDays = math.Round(totalDays/float64(daysCount)*10) / 10
	}

	writeJSON(w, map[string]any{
		"by_assignee":          stats,
		"avg_days_to_complete": avgDays,
		"total_done":           totalDone,
		"recent_completions":   recentCompletions,
	})
}

File: dashboard/lib/queries/tasks.ts
import { getDb } from "../db";
import type { Task } from "../types";

export function listTasks(assignee?: string, status?: string): Task[] {
  let sql = `SELECT id, content, assignee, status, COALESCE(due_date,'') AS due_date, created_at, updated_at FROM tasks`;
  const conditions: string[] = [];
  const args: unknown[] = [];

  if (assignee) {
    conditions.push("LOWER(assignee) = LOWER(?)");
    args.push(assignee);
  }
  if (status) {
    conditions.push("status = ?");
    args.push(status);
  }
  if (conditions.length > 0) {
    sql += " WHERE " + conditions.join(" AND ");
  }
  sql +=
    " ORDER BY CASE WHEN due_date IS NULL OR due_date = '' THEN 1 ELSE 0 END, due_date ASC, created_at DESC";

  return getDb().prepare(sql).all(...args) as Task[];
}

export interface AssigneeStats {
  assignee: string;
  pending: number;
  done: number;
  overdue: number;
}

export function taskStats() {
  const all = listTasks();
  const today = new Date().toISOString().slice(0, 10);
  const byAssignee: Record<string, AssigneeStats> = {};
  let totalDone = 0;
  let totalDays = 0;
  let daysCount = 0;

  for (const t of all) {
    if (!byAssignee[t.assignee]) {
      byAssignee[t.assignee] = {
        assignee: t.assignee,
        pending: 0,
        done: 0,
        overdue: 0,
      };
    }
    const s = byAssignee[t.assignee];

    if (t.status === "pending") {
      s.pending++;
      if (t.due_date && t.due_date < today) {
        s.overdue++;
      }
    } else if (t.status === "done") {
      s.done++;
      totalDone++;
      const created = new Date(t.created_at).getTime();
      const updated = new Date(t.updated_at).getTime();
      if (!isNaN(created) && !isNaN(updated)) {
        const days = Math.max(0, (updated - created) / (1000 * 60 * 60 * 24));
        totalDays += days;
        daysCount++;
      }
    }
  }

  return {
    byAssignee: Object.values(byAssignee),
    totalDone,
    avgDaysToComplete:
      daysCount > 0 ? Math.round((totalDays / daysCount) * 10) / 10 : 0,
    totalPending: all.filter((t) => t.status === "pending").length,
    totalOverdue: all.filter(
      (t) => t.status === "pending" && t.due_date && t.due_date < today,
    ).length,
  };
}
